Clean up highlevel-rest gradle file

tlrx opened the pull request:
This commit cleans up a bit the current Gradle file of the highlevel-rest module and adds a hard check to avoid adding more dependencies.

tlrx committed:
Clean up highlevel-rest gradle file

This commit cleans up a bit the current Gradle file of the highlevel-rest module and adds a hard check to avoid adding more dependencies.

dadoonet reviewed (approved):
LGTM but let's wait for what Luca thinks.

rjernst commented on client/highlevel-rest/build.gradle:
```diff
@@ -36,26 +31,30 @@ dependencies {
-  // Don't check the client's license. We know it.
-  dependencies = project.configurations.runtime.fileCollection {
-    it.group.startsWith('org.elasticsearch') == false
+task checkNoOtherDependencies {
```
This doesn't need to be a task. It can be done at configuration time, see `BuildPlugin.configureConfigurations`. However, I don't see why adding code in the same file that the dependencies are declared is necessary? Why can't this be just a comment next to the two compile dependencies?

tlrx replied:
I'm just afraid that someone blindly copy/paste a new dependency without looking at the comment :/

rjernst replied:
Well make the comment very big, and surround the compile deps. Someone who blindly pastes will also blindly edit.

rjernst commented on client/highlevel-rest/build.gradle:
```diff
@@ -36,26 +31,30 @@ dependencies {
-     configure(BuildPlugin.commonTestConfig(project))
-     classpath = project.test.classpath
-     testClassesDir = project.test.testClassesDir
-     include '**/*IT.class'
```
It seems you removed this, but I would guess you meant to call the method you extracted out in RestTestPlugin?  Instead of importing a class to call a method which creates tasks, create alternate gradle plugins (furthing the composition that is already there). For example, create a IntegTestPlugin, which RestTestPlugin will either extend or wrap (there are other places where we apply `elasticsearch.rest-test` just to have integ tests so having `elasticsearch.integ-test` would be good).

tlrx replied:
Makes sense, thanks!

tlrx committed:
Update after Ryan comments

javanna reviewed (approved):
/me has no idea about this stuff, looks perfect to me

tlrx commented:
Thanks @dadoonet and @javanna.

@rjernst I pushed https://github.com/elastic/elasticsearch/pull/22067/commits/9c4aa0f5bcaf2e327b76a3559da0090e69f94fa5 , is that what you had in mind?

rjernst commented:
> is that what you had in mind?

For the comment around dependencies, yes. For integ test...no. In fact, I'm really not sure what you are trying to accomplish. The only difference I see is the distribution? But you could have set that already and used the rest test plugin. Maybe you can explain why you can't just `apply plugin: elasticsarch.rest-test`?

tlrx committed:
Revert: Update after Ryan comments

tlrx committed:
Update after Ryan comments (2)

tlrx commented:
@rjernst I reverted the previous changes. I'm trying to apply both elasticsearch.build and elasticsearch.rest-test plugins but the latter applies StandaloneTestBasePlugin that duplicates some tasks and config from the BuildPlugin.

I updated this PR to show you what causes issues. It now builds correctly but I'm not sure this is the way to go when you want both releasing a project with artefacts and also have REST tests. Maybe we should have a dedicated `qa` project for high level client REST tests?

tlrx committed:
Revert: Clean up highlevel-rest gradle file

tlrx commented:
I merged this PR in the feature branch so that people are not blocked waiting for it anymore.

I think that the last change in `StandaloneTestBasePlugin` - if not perfect - is at least not intrusive. I'll be happy to have feedback from @rjernst but let no block people for this.
